htlcswitch/link_test.go

```diff
@@ -0,0 +1,626 @@
+	}
+
+	if n.carolChannelLink.Bandwidth() != carolBandwidthBefore {
+		t.Fatal("the bandwidth of carol channel link which handles " +
+			"bob->carol channel should be the same")
+	}
+}
+
+// TestChannelLinkMultiHopDecodeError checks that we send HTLC cancel if
+// decoding of onion blob failed.
+func TestChannelLinkMultiHopDecodeError(t *testing.T) {
+	n := newThreeHopNetwork(t)
+	if err := n.start(); err != nil {
+		t.Fatalf("can't start three hop network: %v", err)
+	}
+	defer n.stop()
+
+	// Replace decode function with another which throws an error.
+	n.carolChannelLink.cfg.DecodeOnion = func(r io.Reader, meta []byte) (
+		HopIterator, error) {
+		return nil, errors.New("some sphinx decode error")
+	}
+
+	carolBandwidthBefore := n.carolChannelLink.Bandwidth()
+	firstBobBandwidthBefore := n.firstBobChannelLink.Bandwidth()
+	secondBobBandwidthBefore := n.secondBobChannelLink.Bandwidth()
+	aliceBandwidthBefore := n.aliceChannelLink.Bandwidth()
+
+	var amount btcutil.Amount = btcutil.SatoshiPerBitcoin
+	errChan, invoice, err := n.makePayment(n.carolServer, amount)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	select {
+	case err := <-errChan:
+		if err == nil ||
+			err.Error() != errors.New(lnwire.SphinxParseError).Error() {
+			t.Fatal("error wasn't received")
+		}
+		break
+	case <-time.After(time.Second):
+		t.Fatal("error was no settled in time")
+	}
+
+	// Wait for Bob to receive the revocation.
+	time.Sleep(100 * time.Millisecond)
+
+	// Check that alice invoice wasn't settled and bandwidth of htlc
+	// links hasn't been changed.
+	if invoice.Terms.Settled {
+		t.Fatal("alice invoice was settled")
+	}
+
+	if n.aliceChannelLink.Bandwidth() != aliceBandwidthBefore {
+		t.Fatal("the bandwidth of alice channel link which handles " +
+			"alice->bob channel should be the same")
+	}
+
+	if n.firstBobChannelLink.Bandwidth() != firstBobBandwidthBefore {
+		t.Fatal("the bandwidth of bob channel link which handles " +
+			"alice->bob channel should be the same")
+	}
+
+	if n.secondBobChannelLink.Bandwidth() != secondBobBandwidthBefore {
+		t.Fatal("the bandwidth of bob channel link which handles " +
+			"bob->carol channel should be the same")
+	}
+
+	if n.carolChannelLink.Bandwidth() != carolBandwidthBefore {
+		t.Fatal("the bandwidth of carol channel link which handles " +
+			"bob->carol channel should be the same")
+	}
+}
+
+// TestChannelLinkSingleHopMessageOrdering test checks ordering of message which
+// flying around between Alice and Bob are correct when Bob sends payments to
+// Alice.
+func TestChannelLinkSingleHopMessageOrdering(t *testing.T) {
+	n := newThreeHopNetwork(t)
+
+	chanPoint := n.aliceChannelLink.ChanID()
+
+	// Append initial channel window revocation messages which occurs after
+	// channel opening.
+	var aliceOrder []lnwire.Message
+	for i := 0; i < lnwallet.InitialRevocationWindow; i++ {
+		aliceOrder = append(aliceOrder, &lnwire.RevokeAndAck{})
+	}
+
+	aliceOrder = append(aliceOrder, []lnwire.Message{
```

Comment:
Is this the order in which the message should have been _sent_ to Alice, _received_ or both?